docs/articles/infrastructure/ci-cd/yocto-cache.rst

```diff
@@ -0,0 +1,101 @@
+Yocto cache
+===========
+
+This chapter explains how to set up a cache for use with yocto. While building
+an image, yocto caches everything it builds and downloads. This is useful since
+it enables incremental builds where one only has to re-build what was changed
+since last build, but also this makes it possible to share this cache between
+builds. When used properly, the cache will dramatically improve build times,
+usually up to 50% faster.
+
+Specifically, we will be talking about the ``downloads`` directory and the
+``sstate-cache`` directory in this chapter. The ``downloads`` directory contains
+all downloaded packages, including those downloaded over svn or git or similar,
+and ``sstate-cache`` contains everything that was built, so basically binaries
+plus some meta-data stating that it has actually been built and what version and
+so on. One reason why the meta-data is useful is because otherwise the system
+might try to reuse binaries for x86 when building for armv7, which wouldn't
+work.
+
+In general, when using a cached build in yocto, there are two options that needs
+to be set: ``SOURCE_MIRROR_URL`` and ``SSTATE_MIRRORS``. The first one points to
+where source packages can be found, the second points to the cached binary
+builds.
+
+.. code-block: bash
+    :caption: site.conf
+
+    INHERIT += "own-mirrors"
+    SOURCE_MIRROR_URL ?= "http://example.org/yocto-cache/downloads/"
+    SSTATE_MIRRORS ?= "file://.* http://example.org/yocto-cache/sstate-cache/PATH"
+
+Notice the PATH variable, which yocto will substitute automatically. Also notice
+that one has to inherit ``own-mirrors`` in yocto for this to work. If we ignore
+the details of how and where the cache comes from (in this case http), this is
+really all that is needed in the build.
+
+Cache over HTTP
+---------------
+Yocto accepts caches that are served over HTTP, one simply inputs the URL as
+seen in the example above. Accessing such a cache is then done automatically.
+But updating the cache can be trickier, but it depends on the setup. The general
+scheme is to simply copy the files from one's build to the cache server. This
+can be done through scp, ftp, or if the cache server is mounted - by simple
+copy.
+
+For the HTTP server, one needs to allow to serve a directory plain and simple.
+For nginx it would look like this:
+
+.. code-block:: none
+    :caption: ``nginx.conf`` on ``example.org``
+
+    http {
+        server {
+            location /yocto-cache/ {
+                alias     /var/www/yocto-cache/;
+                index     index.html index.htm;
+                autoindex on;
+            }
+        }
+    }
+
+When one updates the cache, this means writing to /var/www/yocto-cache in some
+way.
+
+Cache over NFS
+--------------
+Setting up a cache over NFS requires more setup on the server-side, but on the
+other hand, the cache can be kept private from the outside world.
```

Comment:
You can keep the http also either behind a VPN or even easier with a basic-auth password in the URL: `https://user:password@example.com/yocto-cache/` if you use https the URL is send encrypted so no problem having the password there either.

I'd remove the statement about having it private and would rather add something about everyone being able to update the cache or something like that.

Reply:
Fair point!